### PR TITLE
Introduce @public marker to explicitly mark members as public

### LINKED ISSUE #645: add --dunder flag to include documentation for double-underscore methods
#### Problem Description

I would like `pdoc` to produce documentation for double-underscore (dunder) methods in classes. I recognize that `pdoc` currently discourages including docs for these methods, and suggests providing examples instead, but I am using `pdoc` in a tutorial on Python for research scientists and would like it to generate documentation for complex `__eq__` and `__getitem__`/`__setitem__` methods.

#### Proposal

1.  Add a `--dunder` command-line flag which, if given, tells `pdoc` to include documentation for dunder methods _if_ they have docstrings. By default, dunder methods will continue to be ignored (to ensure backward compatibility).
2. If a dunder method does not have a docstring, it will _not_ be included in the generated docs if `--include-undocumented` is provided _unless_ `--dunder` is also provided.

#### Alternatives

Leave things as they are :-)

### PR DESCRIPTION
**Pull request by tmeyier**
Here is an implementation of the @public marker, as proposed in #646. 

* members with "@public" in the docstring will be considered public by the default jinja2 template and shown in the generated API documentation - even if the member name starts with _.
* the definition in `__doc__` has higher precedence than @public
* the "@public" marker is removed from the docstring in the rendered docs. This is done with a simple replace filter. So the @public marker should be added directly before the closing """ of the docstring to avoid additional whitespace in the rendered docs.

closes #645

### CODE REVIEW

File: pdoc/__init__.py [hunk 1]
```diff
@@ -180,9 +180,9 @@ class GoldenRetriever(Dog):
 - If `__all__` is defined in the module, then all identifiers in that list will be considered public.
    No other identifiers will be considered public.
 - If `__all__` is not defined, then pdoc will consider all members public that
-   1. do not start with an underscore,
+   1. do not start with an underscore or have @public in their docstring,
```
Comment (mhils) [inline pdoc/__init__.py:183]:
```suggestion
   1. do not start with an underscore or have `@public` in their docstring,
```

**Review by mhils — COMMENTED** [review]
This looks great, thank you so much for tackling this! 🍰

**Comment by mhils** [comment]
Also sorry @public for the notifications! ❤️  Looks like you're taking this with a bit of humor, which is great. Thank you for your understanding, I'll try to go with `@public` going forward! 😅 🍰

**Comment by public** [comment]
I'm used to it :)

File: test/testdata/visibility.py [hunk 2]
```diff
@@ -12,3 +15,36 @@ def my_func():
...
+def my_other_func():
+    """
+    This is another public method that's shown. It should show without additional
+    whitespace above.
+    """
+
+
+def yet_another_func():
+    """
+    This is yet another public method to give a gauge for the expected amount of
```
[the remark this answers is not in the document]
Comment (tmeyier) [reply test/testdata/visibility.py:48]: I wanted to see if there is any unwanted whitespace after the docstring  in the generated html. The second method is there only to see how much spacing between methods you get regularly. For the automated test suite it's probably not really useful.

File: test/testdata/visibility.py [hunk 3]
```diff
@@ -12,3 +15,36 @@ def my_func():
...
+    """
+    This is a private method that's not shown because its name starts with __.
+    """
+
+
+def __private_func_explicitly_public():
+    """
+    This is a private method that's shown because it is explicitly marked
+    as public.
+    @public"""
```
[the remark this answers is not in the document]
Comment (tmeyier) [reply test/testdata/visibility.py:36]: I don't mind either way. @public at the beginning is also removed without residual white space.

File: test/testdata/visibility.py [hunk 4]
```diff
@@ -8,7 +11,40 @@ class Undocumented:
...
+def public_func():
+    """
+    This is another public method that's shown. It should show without additional
+    whitespace above.
+    """
+
+
+def yet_another_func():
+    """
+    This is yet another public method to give a gauge for the expected amount of
```
Comment (tmeyier) [inline test/testdata/visibility.py:48]:
```suggestion
```

File: test/testdata/visibility.py [hunk 5]
```diff
@@ -1,5 +1,8 @@
 """
-Example where no children should be rendered.
+Example where only the these children should be rendered:
+    * `__private_func_explicitly_public`,
+    * `my_other_func` and
+    * `yet_another_func` should be rendered.
```
Comment (tmeyier) [inline test/testdata/visibility.py:5]:
```suggestion
Example where only the these children should be rendered:
    * `__private_func_explicitly_public` and
    * `public_func` 
```

**Comment by tmeyier** [comment]
The code review functionality in github still is a bit confusing for me. Sorry for the git mess.

**Comment by mhils** [comment]
No worries at all, we'll squash anyways. Thanks again for contributing this, this is really cool! 😃 🍰

**Pull request merged**